### PR TITLE
Customize the training loop for the classifier

### LINKED ISSUE #17: Failure to run training in AWS on Pyspacer Patch Model
## Summary/Issue
Pyspacer training runs do not seem to be finishing, according to ML flow results.

## Impact/Priority
Blocking progress on running with large datasets. For example, I have been unable to complete a training run with all MERMAID annotations (221,983 annotations at last successful training run) with all the CoralNet source annotations that we used in the current MERMAID model (322 sources). 

## Testing Steps
I was able to complete training runs using an "ml.t3.large" Sagemaker instance with all MERMAID annotations plus the first 5 CoralNet sources (451,808 total annotations; run took <1 hour) and with all MERMAID annotations plus first 10 CoralNet sources (556,839 total annotations; run took 1.3 hours) but when I tried to run with MERMAID plus the first 20 CoralNet sources it did not finish even after 9 days (1,109,799 total annotations). As a test, I ran with the same dataset (MERMAID plus first 20 CoralNet sources) but using an XL instance ("ml.t3.xlarge") and it completed in 3.3 hours using that larger instance. I was able to use that XL instance to run with MERMAID annotations plus the first 40 CoralNet Sources (1,775,052 total annotations; run took 5.2 hours) but an even larger dataset (MERMAID plus the first 80 CoralNet sources) failed to complete even after 5 days. Based on this testing, I am guessing that using an even larger instance or changing to a different type of instance (e.g. ml.**m5**.xlarge instead of ml.**t3**.xlarge), but it would be better to determine the underlying issue so we can test with even more data without needing to use huge instances. 

## Request
Could someone please investigate the underlying issue for why large datasets could be failing to complete?

## Recommended Next Steps
I would recommend testing with a larger instance and a different type of instance (e.g. M5) to see if those are actually solutions for larger datasets but I would also recommend looking into the underlying code to see if it is likely a memory issue or something else (see notes from Stephen as well).

Image

### PR DESCRIPTION
**Pull request by gnieuwenhuis**
These changes create a new pyspacer trainer class which inherits from the pyspacer ClassiferTrainer, and defines custom training logic (enabled by the recent pyspacer release 0.14.0).

Custom training logic currently includes:

- added memory logging (system logging in mlflow, with swap space useage as a custom system metric)
- batching of accuracy calculations as well as calibration
- Changes to how calibration is done in order to allow batching (implements the same algorithm, but is being done without the sklearn CalibratedClassifierCV).
- Increased the amount of reserved memory from 2GB to 3GB.
- Updates pyspacer to v 0.14.0

**Note**

These changes were iniitally in https://github.com/data-mermaid/mermaid-classifier/pull/18, which was closed because of some mistaken commits I made and had to awkwardly fix.

Closes: #17

### CODE REVIEW

File: mermaid_classifier/pyspacer/settings.py [hunk 1]
```diff
@@ -5,19 +5,29 @@
...
-    # worked, but 200000 crashed.
-    ref_size = int(rough_available_ram_bytes * (100000 / (6e9)))
+
+    # Reserve 3GB for OS, Python runtime, DuckDB tables, data prep
+    # intermediates (ImageLabels, train_test_split copies), and heap
+    # fragmentation. 
+    base_overhead_bytes = 3e9
+    rough_available_ram_bytes = max(total_ram_bytes - base_overhead_bytes, 0)
+
+    ref_size = int(rough_available_ram_bytes * (50000 / 3e9))
```
Comment (StephenChan) [inline mermaid_classifier/pyspacer/settings.py:32]: I like the upgraded comment over base_overhead_bytes. Right now this ref_size calculation isn't particularly clear though. The old version had a comment over ref_size which explained where the numbers came from: 100000 ref size worked with (8GB - 2GB) RAM, so I aimed for the same ref-size-to-RAM ratio using the detected amount of RAM.

File: mermaid_classifier/pyspacer/settings.py [hunk 2]
```diff
@@ -127,3 +126,11 @@ def set_env_vars_for_packages():
...
             # so that spacer or MLflow sees it.
             os.environ[env_var_name] = var_value
+
+    # Effectively disable PySpacer's ref set size cap.
+    # PySpacer's split_labels() caps the ref set via
+    # config.TRAINING_BATCH_LABEL_COUNT. With batch calibration in
+    # trainer.py, the ref set no longer needs to fit in memory at once,
+    # so we set this to a very large value to let ref_val_ratios alone
+    # determine the ref set size.
+    os.environ['SPACER_TRAINING_BATCH_LABEL_COUNT'] = str(2**31 - 1)
```
Comment (StephenChan) [inline mermaid_classifier/pyspacer/settings.py:138]: Right, I believe you brought up the idea of changing pyspacer to have separate settings for the batch size and ref set cap, or something along those lines. Which I'm on board with, but in terms of a quick fix for mermaid-classifier this is probably the way to go.

File: mermaid_classifier/pyspacer/train.py [hunk 3]
```diff
@@ -595,12 +606,16 @@ def __init__(self, options: DatasetOptions):
...
+
+
         with self.section_profiling("Prep annotations for PySpacer"):
             self.labels = self.prep_annotations_for_pyspacer()
             self.add_training_set_names()
 
         self.set_train_summary_stats()
 
+
+
```
Comment (StephenChan) [inline mermaid_classifier/pyspacer/train.py]: Looks like there are some extra line breaks in this file?

File: mermaid_classifier/pyspacer/train.py [hunk 4]
```diff
@@ -23,12 +23,17 @@
...
 from s3fs.core import S3FileSystem
+from mlflow.tracking.fluent import run_id_to_system_metrics_monitor
+from mermaid_classifier.pyspacer.swap_monitor import SwapMonitor
 from spacer.data_classes import DataLocation, ImageLabels, ValResults
 from spacer.messages import TrainClassifierMsg
 from spacer.storage import load_classifier
-from spacer.tasks import train_classifier
 from spacer.task_utils import preprocess_labels, SplitMode
 
+from spacer.tasks import train_classifier as spacer_train_classifier
```
Comment (StephenChan) [inline mermaid_classifier/pyspacer/train.py]: If I recall correctly, did this "as name" import come from the previous version of PR #18, where you were defining your own train_classifier() function to get past pyspacer limitations? If so, then just wanted to point out that this alternate name isn't strictly needed anymore, though if the alternate name still helps you then feel free to keep it.

File: tests/pyspacer/test_trainer.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        clf_standard.fit(X_ref, y_ref)
+
+        # Batched calibration (new approach) via mock ImageLabels
+        mock_labels = _make_mock_labels(X_ref, y_ref, batch_size=100)
+        trainer = MermaidTrainer(batch_size=100)
+        clf_batched = trainer._calibrate_in_batches(clf, mock_labels)
+
+        return clf_standard, clf_batched, X_ref
+
+    def test_sgd_multiclass_calibration_equivalence(self):
```
Comment (StephenChan) [inline tests/pyspacer/test_trainer.py:98]: My local machine (Linux AMD for what it's worth) has trouble with this test specifically. I tried about 7 different values for the random seed (e.g. changing from 42 to 43, 44, 45...) before I got this to pass. Meanwhile all of the other tests passed every time. I don't really have a sense of what tolerance level needs to fail before we should be concerned, but just wondering about your experience with the consistency of these tests?

**Comment by StephenChan** [comment]
I went through this PR and the 'Speed up training' PR, leaving comments and testing them myself. I saw the third PR (metrics system) was still in draft, but let me know when it's ready to look at too.

- Nice to see what the System metrics tab is for! Really feels like it's making good use of MLflow now.
- Already commented on the batching during calibration in issue 17, so I'll just [link that comment](https://github.com/data-mermaid/mermaid-classifier/issues/17#issuecomment-3970282448) for reference. Of course the adjacent code is nicer now due to the ClassifierTrainer rework, so that's good to see.

File: mermaid_classifier/pyspacer/settings.py (hunk 1 above)
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/settings.py:32]: hmm, right, I changed the numbers while holding the ratio the same and losing the comment. I've re-added the comment along with the new ratio values

File: mermaid_classifier/pyspacer/train.py (hunk 3 above)
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/train.py]: Yeah, this was related to my IDE settings, reverted.

File: mermaid_classifier/pyspacer/train.py (hunk 4 above)
Comment (gnieuwenhuis) [reply mermaid_classifier/pyspacer/train.py]: definitely just overlooked it - reverted to minimize changes.

File: tests/pyspacer/test_trainer.py (hunk 5 above)
Comment (gnieuwenhuis) [reply tests/pyspacer/test_trainer.py:98]: These tests were pretty flaky - I adjusted them so that they should be more robust (instead of completely flat and random data, use separable Gaussian clusters), as well as increasing tolerance to ensure they run properly in different environments.

I also added a github action to run the tests on linux architecture (because I am developing on mac, but training runs in a linux env). This should avoid this kind of problem in the future as well.

File: mermaid_classifier/pyspacer/settings.py (hunk 1 above)
Comment (StephenChan) [reply mermaid_classifier/pyspacer/settings.py:32]: OK, re-added comment looks good.

File: tests/pyspacer/test_trainer.py (hunk 5 above)
Comment (StephenChan) [reply tests/pyspacer/test_trainer.py:98]: Great work on this! The tests now pass consistently for me (once again trying about 7 different random seeds) and it's also more satisfying to have per-class centroids, since that's closer to the kind of real-world data we're expecting. The comments regarding the tolerances are helpful, and the CI action's great to have too.

**Review by StephenChan — APPROVED** [review]

**Comment by StephenChan** [comment]
> Sorry for the time it took - I have been largely focussed on experimentation, and didn't want to go through the review cycle on many of these until I was sure the changes were justified.

No worries, I saw the other draft PRs going around and figured it was something like that. I think you've resolved all of the points I brought up in this PR by now, so it looks good to go. I'll follow up on PR 26 later probably, and 27 next week.
